dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -4323,7 +4323,10 @@ RtpsUdpDataLink::RtpsReader::deliver_held_data(const RepoId& src)
     const SequenceNumber ca = wi->second->recvd_.cumulative_ack();
     const WriterInfo::HeldMap::iterator end = wi->second->held_.upper_bound(ca);
     for (WriterInfo::HeldMap::iterator it = wi->second->held_.begin(); it != end; /*increment in loop body*/) {
-      to_deliver.push_back(it->second);
+      const DCPS::SystemTimePoint source_time(ACE_Time_Value(it->second.header_.source_timestamp_sec_, it->second.header_.source_timestamp_nanosec_ / 1000));
```

Comment:
This seems like something that should be a getter in `DataSampleHeader`. This along with a setter that would be used `deliver_sample_i`.